src/Common/Core/Test/Services/ServiceManagerTests.cs

```diff
@@ -42,26 +36,74 @@ public class ServiceManagerTests {
         }
 
         [Test]
-        public void DoubleAdd() {
-            _serviceManager.AddService<I1>(new C1());
-            
-            Action a = () => _serviceManager.AddService<I1>(new C1());
-            a.ShouldThrow<InvalidOperationException>();
-        }
-
-        [Test]
         public void AddRemove() {
             var s = new C1();
             _serviceManager.AddService(s);
             _serviceManager.RemoveService(s);
             _serviceManager.GetService<I1>().Should().BeNull();
 
-            _serviceManager.AddService(s as I1);
+            _serviceManager.AddService((I1) s);
             _serviceManager.RemoveService<I1>();
             _serviceManager.GetService<I1>().Should().BeNull();
         }
 
         [Test]
+        public void AddRemoveDerived() {
+            var d = new Derived(_serviceManager);
+            _serviceManager.GetService<Derived>().Should().NotBeNull();
+            _serviceManager.GetService<Base>().Should().NotBeNull();
+            _serviceManager.GetService<I1>().Should().NotBeNull();
+            _serviceManager.GetService<I2>().Should().NotBeNull();
+
+            d.Dispose();
+            _serviceManager.GetService<Derived>().Should().BeNull();
+            _serviceManager.GetService<Base>().Should().BeNull();
+            _serviceManager.GetService<I1>().Should().BeNull();
+            _serviceManager.GetService<I2>().Should().BeNull();
+        }
+
+        [Test]
+        public void AddRemoveLazy01() {
+            _serviceManager.AddService<I2, Derived>();
+            _serviceManager.RemoveService<I2>();
+
+            _serviceManager.GetService<I1>().Should().BeNull();
+            _serviceManager.GetService<I2>().Should().BeNull();
+        }
+
+        [Test]
+        public void AddRemoveLazy02() {
+            _serviceManager.AddService<I2, Derived>();
+            _serviceManager.RemoveService<I1>();
+
+            _serviceManager.GetService<I1>().Should().BeNull();
+            _serviceManager.GetService<I2>().Should().BeNull();
+        }
+
+        [Test]
+        public void AddRemoveLazy03() {
+            _serviceManager.AddService<I2, Derived>();
+            _serviceManager.RemoveService<Base>();
```

Comment:
The example that I've provided is actually wrong - `_services.Remove(_service)` will work because it checks for instance type, not for the type of variable.
Generally, we have two different cases:
- Service is added as instance or as a concrete type, which is backward compatible with editor code. For that, we have `_serviceManager.AddService(service)`, `_serviceManager.Remove(service)` and `_serviceManager.AddService<ServiceType>()`. These services can be found by any interface they implement as well as type itself.
- Service is especially added by its interface or some other key. For that, we call `_serviceManager.AddService(service, typeof(InterfaceType))` and `_serviceManager.AddService<InterfaceType, ServiceType>()`. This provides better code decoupling and testability and shouldn't provide access to service by anything except the specified interface.

`IServiceManager.RemoveService<T>` doesn't fall into any of these cases. Since we don't use it right now, maybe we can just remove it?